pkg/machineoutput/event_logging.go

```diff
@@ -0,0 +1,258 @@
+}
+
+// DevFileCommandExecutionComplete outputs the provided event as JSON to the console.
+func (c *ConsoleMachineEventLoggingClient) DevFileCommandExecutionComplete(commandID string, componentName string, commandLine string, groupKind string, timestamp string, errorVal error) {
+
+	errorStr := ""
+
+	if errorVal != nil {
+		errorStr = errorVal.Error()
+	}
+
+	json := MachineEventWrapper{
+		DevFileCommandExecutionComplete: &DevFileCommandExecutionComplete{
+			CommandID:        commandID,
+			ComponentName:    componentName,
+			CommandLine:      commandLine,
+			GroupKind:        groupKind,
+			AbstractLogEvent: AbstractLogEvent{Timestamp: timestamp},
+			Error:            errorStr,
+		},
+	}
+
+	OutputSuccessUnindented(json)
+}
+
+// CreateContainerOutputWriter returns an io.PipeWriter for which the devfile command/action process output should be
+// written (for example by passing the io.PipeWriter to exec.ExecuteCommand), and a channel for communicating when the last data
+// has been received on the reader.
+//
+// All text written to the returned object will be output as a log text event.
+// Returned channels will each contain a single nil entry once the underlying reader has closed.
+func (c *ConsoleMachineEventLoggingClient) CreateContainerOutputWriter() (*io.PipeWriter, chan interface{}, *io.PipeWriter, chan interface{}) {
+
+	stdoutWriter, stdoutChannel := createWriterAndChannel(false)
+	stderrWriter, stderrChannel := createWriterAndChannel(true)
+
+	return stdoutWriter, stdoutChannel, stderrWriter, stderrChannel
+
+}
+
+func createWriterAndChannel(stderr bool) (*io.PipeWriter, chan interface{}) {
+	reader, writer := io.Pipe()
+
+	closeChannel := make(chan interface{})
+
+	stream := "stdout"
+	if stderr {
+		stream = "stderr"
+	}
+
+	go func() {
+
+		bufReader := bufio.NewReader(reader)
+		for {
+			line, _, err := bufReader.ReadLine()
+			if err != nil {
+				klog.V(4).Infof("Unexpected error on reading container output reader: %v", err)
+				break
+			}
+
+			json := MachineEventWrapper{
+				LogText: &LogText{
+					AbstractLogEvent: AbstractLogEvent{Timestamp: TimestampNow()},
+					Text:             string(line),
+					Stream:           stream,
+				},
+			}
+			OutputSuccessUnindented(json)
+		}
+
+		// Output a single nil event on the channel to inform that the last line of text has been
+		// received from the writer.
+		closeChannel <- nil
+	}()
+
+	return writer, closeChannel
+}
+
+// ReportError ignores the provided event.
+func (c *ConsoleMachineEventLoggingClient) ReportError(errorVal error, timestamp string) {
+	json := MachineEventWrapper{
+		ReportError: &ReportError{
+			Error:            errorVal.Error(),
+			AbstractLogEvent: AbstractLogEvent{Timestamp: timestamp},
+		},
+	}
+
+	OutputSuccessUnindented(json)
+}
+
+// GetEntry will return the JSON event parsed from a single line of '-o json' machine readable console output.
+func (w MachineEventWrapper) GetEntry() (MachineEventLogEntry, error) {
```

Comment:
Unit tests for `GetEntry`?

Reply:
This is likewise only used for test purposes (I will add a comment indicating this), and so it likely doesn't make sense to unit test it (and likewise since the implementation is so simple).

Reply:
Yeah, a comment indicating it's just for test purposes is good for me